app/code/Magento/GoogleGtag/Block/Ga.php

```diff
@@ -0,0 +1,200 @@
+     *
+     * @param string $measurementId
+     * @return array
+     */
+    public function getPageTrackingData($measurementId): array
+    {
+        return [
+            'optPageUrl' => $this->getOptPageUrl(),
+            'measurementId' => $this->escapeHtmlAttr($measurementId, false)
+        ];
+    }
+
+    /**
+     * Return information about order and items for GA tracking.
+     *
+     * @link https://developers.google.com/analytics/devguides/collection/ga4/ecommerce#purchase
+     * @link https://developers.google.com/gtagjs/reference/ga4-events#purchase
+     * @link https://developers.google.com/analytics/devguides/collection/gtagjs/enhanced-ecommerce#product-data
+     * @link https://developers.google.com/gtagjs/reference/event#purchase
+     *
+     * @return array
+     */
+    public function getOrdersTrackingData(): array
+    {
+        $result = [];
+        $orderIds = $this->getOrderIds();
+        if (empty($orderIds) || !is_array($orderIds)) {
+            return $result;
+        }
+        $this->searchCriteriaBuilder->addFilter(
+            'entity_id',
+            $orderIds,
+            'in'
+        );
+        $collection = $this->orderRepository->getList($this->searchCriteriaBuilder->create());
+
+        foreach ($collection->getItems() as $order) {
+            foreach ($order->getAllVisibleItems() as $item) {
+                $result['products'][] = [
+                    'item_id' => $this->escapeJsQuote($item->getSku()),
+                    'item_name' =>  $this->escapeJsQuote($item->getName()),
+                    'price' => number_format((float) $item->getPrice(), 2),
+                    'quantity' => (int)$item->getQtyOrdered(),
+                ];
+            }
+            $result['orders'][] = [
+                'transaction_id' =>  $order->getIncrementId(),
+                'affiliation' => $this->escapeJsQuote($this->_storeManager->getStore()->getFrontendName()),
+                'value' => number_format((float) $order->getGrandTotal(), 2),
+                'tax' => number_format((float) $order->getTaxAmount(), 2),
+                'shipping' => number_format((float) $order->getShippingAmount(), 2),
+            ];
+            $result['currency'] = $order->getOrderCurrencyCode();
+        }
+        return $result;
+    }
+
+    /**
+     * Return page url for tracking.
+     *
+     * @return string
+     */
+    private function getOptPageUrl(): string
+    {
+        $optPageURL = '';
+        $pageName = $this->getPageName() !== null ? trim($this->getPageName()) : '';
+        if ($pageName && substr($pageName, 0, 1) === '/' && strlen($pageName) > 1) {
+            $optPageURL = ", '" . $this->escapeHtmlAttr($pageName, false) . "'";
+        }
+        return $optPageURL;
+    }
+
+    /**
+     * Provide analytics events data
+     *
+     * @return bool|string
+     */
+    public function getAnalyticsData()
```

Comment:
This method seems to be only one needing to be public in this block. All other methods seem to be okay being private. Please double check this and reduce visibility scope for as much methods as you can.

Additionally, please consider changing this block to a view model. To me it makes much more sense in this case.